core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Manager.java

```diff
@@ -0,0 +1,220 @@
+        // fetch a token using the client credentials flow
+        String credential = credentials.get(OAuth2Properties.CREDENTIAL);
+        return sessionCache.cachedSession(
+            keyFunc.apply(OAuth2Properties.CREDENTIAL),
+            () ->
+                OAuth2Util.AuthSession.fromCredential(
+                    client, refreshExecutor(), credential, parent));
+      }
+
+      for (String tokenType : TOKEN_PREFERENCE_ORDER) {
+        if (credentials.containsKey(tokenType)) {
+          // exchange the token for an access token using the token exchange flow
+          String token = credentials.get(tokenType);
+          return sessionCache.cachedSession(
+              keyFunc.apply(tokenType),
+              () ->
+                  OAuth2Util.AuthSession.fromTokenExchange(
+                      client, refreshExecutor(), token, tokenType, parent));
+        }
+      }
+    }
+    return parent;
+  }
+
+  private static AuthConfig createConfig(Map<String, String> props) {
+    String scope = props.getOrDefault(OAuth2Properties.SCOPE, OAuth2Properties.CATALOG_SCOPE);
+    Map<String, String> optionalOAuthParams = OAuth2Util.buildOptionalParam(props);
+    String oauth2ServerUri =
+        props.getOrDefault(OAuth2Properties.OAUTH2_SERVER_URI, ResourcePaths.tokens());
+    boolean keepRefreshed =
+        PropertyUtil.propertyAsBoolean(
+            props,
+            OAuth2Properties.TOKEN_REFRESH_ENABLED,
+            OAuth2Properties.TOKEN_REFRESH_ENABLED_DEFAULT);
+    return AuthConfig.builder()
+        .credential(props.get(OAuth2Properties.CREDENTIAL))
+        .token(props.get(OAuth2Properties.TOKEN))
+        .scope(scope)
+        .oauth2ServerUri(oauth2ServerUri)
+        .optionalOAuthParams(optionalOAuthParams)
+        .keepRefreshed(keepRefreshed)
+        .expiresAtMillis(expiresAtMillis(props))
+        .build();
+  }
+
+  private static Duration sessionTimeout(Map<String, String> props) {
+    long expirationIntervalMs =
+        PropertyUtil.propertyAsLong(
+            props,
+            CatalogProperties.AUTH_SESSION_TIMEOUT_MS,
+            CatalogProperties.AUTH_SESSION_TIMEOUT_MS_DEFAULT);
+    return Duration.ofMillis(expirationIntervalMs);
+  }
+
+  private static Long expiresAtMillis(Map<String, String> props) {
+    Long expiresInMillis = null;
+    if (props.containsKey(OAuth2Properties.TOKEN)) {
+      expiresInMillis = OAuth2Util.expiresAtMillis(props.get(OAuth2Properties.TOKEN));
+    }
+    if (expiresInMillis == null) {
+      if (props.containsKey(OAuth2Properties.TOKEN_EXPIRES_IN_MS)) {
+        long millis =
+            PropertyUtil.propertyAsLong(
+                props,
+                OAuth2Properties.TOKEN_EXPIRES_IN_MS,
+                OAuth2Properties.TOKEN_EXPIRES_IN_MS_DEFAULT);
+        expiresInMillis = System.currentTimeMillis() + millis;
+      }
+    }
+    return expiresInMillis;
+  }
+
+  private static Map<String, String> configHeaders(Map<String, String> props) {
```

Comment:
These aren't auth headers, so it's awkward that they're being handled by the AuthManager.  This also makes it inconsistent with other auth managers.

Reply:
Fixed in fbc3555.